File: laser_maxrange_republish/scripts/fov_narrower.py
#!/usr/bin/env python
"""
   Author :  Samuel Chieng Kien Ho
   Function :  (1)
               (2)
"""

import rospy
import numpy as np

from sensor_msgs.msg import LaserScan


class FOVNarrower():
    def __init__(self):
        # Define Adjustable Parameters
        # - Scan Region (Round)
        self.ang_min = float(rospy.get_param("~ang_min"))    # [deg]
        self.ang_max = float(rospy.get_param("~ang_max"))    # [deg]
        self.intensity = bool(rospy.get_param("~intensity"))    # [1 or 0]

        # Subscriber
        self.scan_sub = rospy.Subscriber("scan", LaserScan, self.scanCB, queue_size=1)

        # Publisher
        self.scan_pub = rospy.Publisher("scan_filter", LaserScan, queue_size=1)

    def scanCB(self, msg):
        range_list = list(msg.ranges)
        intensity_list = list(msg.intensities)
        for i in xrange(len(msg.ranges)):
            current_angle = msg.angle_min + msg.angle_increment*i
            # Delete the first element
            if (current_angle < np.deg2rad(self.ang_min)):
                del range_list[0]
                del intensity_list[0]
            # Delete the last element
            elif (current_angle > np.deg2rad(self.ang_max)):
                del range_list[-1]
                del intensity_list[-1]
        new_scan = LaserScan()
        new_scan.header = msg.header
        new_scan.angle_min = np.deg2rad(self.ang_min)
        new_scan.angle_max = np.deg2rad(self.ang_max)
        new_scan.angle_increment = msg.angle_increment
        new_scan.time_increment = msg.time_increment
        new_scan.scan_time = msg.scan_time
        new_scan.range_min = msg.range_min
        new_scan.range_max = msg.range_max
        new_scan.ranges = range_list
        if(self.intensity == False):
            new_scan.intensities = []
        else:
            new_scan.intensities = intensity_list
        self.scan_pub.publish(new_scan)



if __name__=="__main__":
    rospy.init_node("fov_narrower")
    FOVNarrower()
    rospy.spin()

File: laser_maxrange_republish/src/fov_narrower.cpp
/*   Author : Samuel Chieng Kien Ho   */

#include <math.h>
#include "ros/ros.h"
#include "std_msgs/String.h"
#include "sensor_msgs/LaserScan.h"

ros::Publisher *scan_pub_;
double ang_min, ang_max;
bool intensity;

inline double to_radians(double degrees) {
    return degrees * (M_PI / 180.0);
}

// "Crop" the Field of View (FOV) of the LaserScan by appointing both minimum and maximum angle
sensor_msgs::LaserScan reduce_fov(sensor_msgs::LaserScanConstPtr msg, double ang_min, double ang_max, bool allow_intensity){
    sensor_msgs::LaserScan ls = *msg;

    for(int i=0; i<msg->ranges.size(); i++){
        double ang = msg->angle_min + msg->angle_increment*i;
        if(ang < to_radians(ang_min)){
            ls.ranges.erase(ls.ranges.begin());
            ls.intensities.erase(ls.intensities.begin());
        }else if(ang > to_radians(ang_max)){
            ls.ranges.erase(ls.ranges.end()-1);
            ls.intensities.erase(ls.intensities.end()-1);
        }
    }
    if(allow_intensity == false){
        ls.intensities.resize(0);
    }
    ls.angle_min = to_radians(ang_min);
    ls.angle_max = to_radians(ang_max);

    return ls;
}

void scanCB(const sensor_msgs::LaserScanConstPtr msg){
    sensor_msgs::LaserScan ls = reduce_fov(msg, ang_min, ang_max, intensity);
    scan_pub_->publish(ls);
}



int main(int argc, char **argv){
    ros::init(argc, argv, "laser_fov_narrower");
    ros::NodeHandle nh;
    ros::NodeHandle priv_nh("~");

    priv_nh.param("ang_min", ang_min, -181.0);
    priv_nh.param("ang_max", ang_max, 180.0);
    priv_nh.param("intensity", intensity);

    ros::Publisher scan_pub = nh.advertise<sensor_msgs::LaserScan>("scan_filter", 1);
    scan_pub_ = &scan_pub;
    ros::Subscriber scan_sub = nh.subscribe("scan", 1, scanCB);

    ros::spin();
    return 0;
}
